CORE-15351: log trace span to separate log file

ZawarR3 opened the pull request:
When deploying Corda with the combined worker, if the following command line argument is provided: `  --send-trace-to=http://localhost:9411 `, the trace data (trace spans) is sent over to the ZipKin Server, as a valid host URL has been provided, however if no URL is provided the tracing data is not generated nor logged anywhere. With this change, when no ZipKin host URL is provided the trace spans will be logged locally into a file called `corda-trace-spans.log`. Developers have the flexibility to configure logging behaviour through the log4j2.xml file. They can choose to enable or disable this local logging with or without ZipKin and can also opt to direct logs to the console instead of log files if they prefer.

ZawarAliR3 committed:
Tracer now used even without zipkin URL provision

ZawarAliR3 committed:
Trace spans now logged to seperate log file

josephzunigadaly commented on libs/rest/rest-server-impl/build.gradle:
```diff
@@ -17,7 +17,8 @@ dependencies {
     implementation project(':libs:rest:json-serialization')
     implementation project(':libs:metrics')
-    implementation project(':libs:tracing')
+    implementation project(':libs:' +
+            'tracing')
```
This looks like an accidental edit. Please can we undo this change?

josephzunigadaly commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -70,20 +70,26 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String, samp
-        reporters.add(spanAsyncReporter)
+            // The console reporter is useful when debugging test runs on the combined worker.
+            // uncomment it to enable it.
+            val reporters = mutableListOf<Reporter<Span>>(Reporter.CONSOLE)
```
`Reporter.CONSOLE` should still be commented out

```suggestion
            val reporters = mutableListOf<Reporter<Span>>(/*Reporter.CONSOLE*/)
```

josephzunigadaly commented on osgi-framework-bootstrap/src/main/resources/log4j2.xml:
```diff
@@ -60,6 +84,10 @@
             <AppenderRef ref="Ledger-Detailed"/>
         </logger>
 
+        <logger name="brave.Tracer" level="info">
```
I think the default level should be "off" for this so that it can be enabled when wanted but normally is switched off.

owenstanford replied:
We should exclude `brave.tracer` from the other loggers so we only get the trace output in the trace log file.

ZawarR3 replied:
I agree, currently working on it.

josephzunigadaly commented on osgi-framework-bootstrap/src/main/resources/log4j2.xml:
```diff
@@ -50,6 +50,30 @@
+
+        <RollingFile name="Trace-Spans"
+                     fileName="logs/corda-trace-spans.log"
+                     filePattern="logs/corda-trace-spans.%d{MM-dd-yyyy}.%i.log"
```
I think `yyyy-MM-dd` is a better date format because it is always sortable.

ZawarAliR3 committed:
Made the following changes:
- console logging is disabled by default
- local trace span logging disabaled by default
- local trace span logging date format changed

ZawarAliR3 committed:
Trace spans now log solely to their separate log file

ZawarAliR3 committed:
Added further configuration to trace span log destination:
- trace spans can now be logged solely to zipkin
- trace spans can now be logged solely to local log files
- trace spans can now be logged solely to console
- trace spans can now be logged to all/some or none

ZawarAliR3 committed:
Removed console reporter in BraveTracingService.kt
- since it console logging can be configured from log4j
- also, set new local trace span logging to off by default

owenstanford commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -70,27 +71,38 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String?, sam
 
-            tracingBuilder.addSpanHandler(spanHandler)
+        override fun report(span: Span) {
+                logger.info(span.toString())
```
We should check to see if logging is enabled before logging to save the cost of the  span.toString()
```if (!logger.isLoggable(Level.INFO)) return ;```

ZawarR3 replied:
is this okay: `       
 override fun report(span: Span) {
            if (logger.isLoggable(Level.INFO)) {
                logger.info(span.toString())
            }
        }`

owenstanford replied:
yes

owenstanford reviewed (commented):
We should be able to delete the `NoopTracingService` class now.

ZawarR3 commented on osgi-framework-bootstrap/src/main/resources/log4j2.xml:
```diff
@@ -10,6 +10,9 @@
             <PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} %X - %msg%n"/>
+            <Filters>
+                <RegexFilter regex=".*brave\.Tracer.*" onMatch="DENY" onMismatch="NEUTRAL"/>
+            </Filters>
```
I forgot to remove this filter, it's not needed. I'll remove it on the next commit.

ZawarAliR3 committed:
Added check to ensure logging only occurs if enabled, to save cost

simon-johnson-r3 commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -22,16 +22,18 @@ import zipkin2.reporter.AsyncReporter
 import zipkin2.reporter.brave.ZipkinSpanHandler
 import zipkin2.reporter.urlconnection.URLConnectionSender
-import java.util.Stack
+import java.util.*
```
I'm not sure we allow wildcard imports, although admittedly the Kotlin standard don't explicitly prevent it.

simon-johnson-r3 commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -70,23 +72,42 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String, samp
 
-        val spanHandler = ZipkinSpanHandler.create(reporter)
+        //Establish zipkin connection iff url host is provided and create respective reporter
+        if (zipkinHost != null){
```
there should be a space between ) and { here, I would run this through the intellij formatter to save you some time and prevent these kinds of formatting comments in PRs generally

simon-johnson-r3 commented on libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:
```diff
@@ -70,23 +72,42 @@ internal class BraveTracingService(serviceName: String, zipkinHost: String, samp
+        }
+
+        // LogReporter will report trace spans to local log files
+        val logReporter = LogReporter()
```
Kotlin can be quite a terse language and generally developers try to embrace that. Here I would suggest
```
reporters.add(LogReporter())
```
It's not a competition to reduce the number of code lines, but removal of variables which are only used once and which call a trivial constructor illustrates to the reader this is the only point of use, and it's one thing Kotlin is quite good at promoting IMHO

ZawarAliR3 committed:
Removed wildcard import and made formatting adjustments

ZawarR3 committed:
Merge branch 'release/os/5.1' into CORE-15351/log_trace_span_to_separate_log_file
